tests/test_rois.py

```diff
@@ -1,209 +1,239 @@
-        8) 3 x 3 object with certain pixels below threshold, dilated twice
+@pytest.mark.parametrize("mask, full, shape, expected", [
+    (
+        np.array([
+            [0.0, 0.0, 0.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0]]),
+        False,
+        None,
+        np.array([
+            [1.0, 1.0],
+            [1.0, 1.0]])),
+    (
+        np.array([
+            [0.0, 0.0, 0.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0]]),
+        True,
+        None,
+        np.array([
+            [0.0, 0.0, 0.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0]]),
+        ),
+    (
+        np.array([
+            [0.0, 0.0, 0.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0]]),
+        False,
+        (5, 5),
+        np.array([
+            [0.0, 0.0, 0.0, 0.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0, 0.0]]),
+        ),
+    (
+        np.array([
+            [0.0, 0.0, 0.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 1.0, 1.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0]]),
+        False,
+        (6, 6),
+        np.array([
+            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+            [0.0, 0.0, 1.0, 1.0, 0.0, 0.0],
+            [0.0, 0.0, 1.0, 1.0, 0.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]),
+        ),
+    ])
+def test_roi_generate_mask(mask, shape, full, expected):
+    coo = coo_matrix(mask)
+    roi = roi_module.ROI(
+            coo.row,
+            coo.col,
+            coo.data,
+            image_shape=(4, 4),
+            experiment_id=1234,
+            roi_id=4567)
+    roi_mask = roi.generate_ROI_mask(shape=shape, full=full)
+    assert np.all(roi_mask == expected)
+
+
+@pytest.mark.parametrize(
+        ("weighted", "expected", "athresh", "cthresh", "full"),
+        [
+            (
+                np.array([
+                    [0.0, 0.0, 0.0, 0.0, 0.0],
+                    [0.0, 0.5, 1.0, 0.0, 0.0],
+                    [0.0, 2.0, 2.0, 1.5, 0.0],
+                    [2.0, 1.0, 1.5, 0.5, 0.0],
+                    [0.0, 0.0, 0.0, 0.0, 0.0]
+                    ]),
+                np.array([
+                    [0, 0, 0, 0, 0],
+                    [0, 0, 1, 0, 0],
+                    [0, 1, 0, 1, 0],
+                    [1, 1, 1, 0, 0],
+                    [0, 0, 0, 0, 0]
+                    ]),
+                0.7,
+                None,
+                True)])
+def test_roi_generate_outline(weighted, full, expected, athresh, cthresh):
+    """this test is not exhaustive
```

Comment:
I would consider adding a test case where cv contours finds multiple contours for one ROI, this should be handled in your function by np.concatinate(contours) but I think it's an important test to have.

Reply:
I'm not sure what you mean here. It is only ever passed 1 ROI. That's what the concatenate is for.

Reply:
I had cases where one ROI was segmented into multiple contours, ie: one ROI had four or five indexes in the upper level list as cv2 said they weren't part of the same shape. This was found during testing on the ROIs table. The concatenate should catch this but for complete test coverage I think it's a good idea to include so we have record. If we're adding tests later however it can probably wait but it's a good test case.

Reply:
I expect the multiple elements of the contour list. That's why I added a `np.concatenate()`
There's nothing wrong here with us getting multiple contours back.

Reply:
I agree there's nothing wrong, and i agree with the concatenate. It was an bug I earlier not using the concatenate and thought it would be a good test case to keep in there. But if we're aware of it letting it pass is fine with me.

Reply:
I agree with Isaak here. An example is if the soma is missing from the ROI, so there's a large "hole" in the middle. You'd expect 2 outlines -- one for the hole and one for the outside.

Reply:
I see, you want the test case where you explicitly see multiple contours. We're quite often getting multiple contours even in a not-so-obvious example as that. But, sure.

Reply:
working on adding this

Reply:
done